File: typescript/src/instrument.ts
import type { RevalexClient } from "./client.js";

/* ────────────────────────────────────────────────────────────────
 * Auto-instrumentation — wrap an Anthropic or OpenAI client so
 * every LLM call is traced automatically. Zero dependencies:
 * duck-typed against the providers' stable response shapes.
 *
 * Contract: tracing NEVER changes behavior — the underlying call's
 * result/error passes through untouched; capture failures are
 * swallowed (fail-open, same as RevalexClient).
 * ──────────────────────────────────────────────────────────────── */

export interface InstrumentOptions {
  /** Trace name; defaults to the provider method name. */
  name?: string;
  /** Capture message content (input/output). Default true — disable for privacy. */
  captureContent?: boolean;
}

const MAX_CAPTURE = 20_000;

function short(v: unknown): string | undefined {
  try {
    const s = typeof v === "string" ? v : JSON.stringify(v);
    return s.length > MAX_CAPTURE ? `${s.slice(0, MAX_CAPTURE)}…` : s;
  } catch {
    return undefined;
  }
}

/* ── Anthropic ─────────────────────────────────────────────────── */

interface AnthropicLike {
  messages: { create: (params: any, ...rest: any[]) => Promise<any> };
}

/**
 * Wrap an Anthropic SDK client: `instrumentAnthropic(anthropic, revalex)`.
 * Every `messages.create` call is traced (model, tokens, latency,
 * optionally content). Streaming calls are traced with latency +
 * model only (no output capture). Returns the same client.
 */
export function instrumentAnthropic<T extends AnthropicLike>(
  client: T,
  revalex: RevalexClient,
  opts: InstrumentOptions = {},
): T {
  const original = client.messages.create.bind(client.messages);
  const capture = opts.captureContent ?? true;

  client.messages.create = ((params: any, ...rest: any[]) => {
    const startedAt = new Date().toISOString();
    const t0 = Date.now();
    let call: Promise<any>;
    try {
      call = original(params, ...rest);
    } catch (err) {
      recordFailure(err);
      throw err;
    }
    // Observe on a SEPARATE chain and return the ORIGINAL object: the
    // provider's create() returns an APIPromise whose .withResponse() /
    // .asResponse() callers broke when we replaced it with a plain async
    // function. Our .then never alters what the caller receives, and the
    // rejection stays theirs to handle (we observe, not consume).
    void Promise.resolve(call).then(recordSuccess, recordFailure);
    return call;

    function recordSuccess(result: any): void {
      try {
        const streaming = params?.stream === true;
        const blocks: any[] = streaming ? [] : (result?.content ?? []);
        const textOut = streaming
          ? undefined
          : blocks
              .filter((b: any) => b?.type === "text")
              .map((b: any) => b.text)
              .join("\n");
        // Capture the model's INTENDED tool calls. These blocks were being
        // dropped, so an auto-instrumented trace had no steps at all and the
        // dashboard's "what it planned" section never rendered. They are the
        // agent's intent, not observed evidence — the provenance check
        // deliberately ignores llm_call content for exactly that reason.
        const planned = blocks
          .filter((b: any) => b?.type === "tool_use")
          .slice(0, 50)
          .map((b: any, i: number) => ({
            index: i,
            type: "llm_call" as const,
            name: String(b?.name ?? "tool_use"),
            toolName: b?.name ? String(b.name) : undefined,
            input: capture ? short(b?.input) : undefined,
            status: "ok" as const,
          }));
        revalex.trace({
          type: "llm_call",
          name: opts.name ?? "anthropic.messages.create",
          model: result?.model ?? params?.model,
          input: capture ? short(params?.messages) : undefined,
          output: capture && textOut ? short(textOut) : undefined,
          steps: planned.length > 0 ? planned : undefined,
          tokensIn: result?.usage?.input_tokens,
          tokensOut: result?.usage?.output_tokens,
          latencyMs: Date.now() - t0,
          status: "ok",
          metadata: streaming ? { streaming: true } : undefined,
          startedAt,
          endedAt: new Date().toISOString(),
        });
      } catch {
        /* capture must never break the app */
      }
    }

    function recordFailure(err: unknown): void {
      try {
        revalex.trace({
          type: "llm_call",
          name: opts.name ?? "anthropic.messages.create",
          model: params?.model,
          input: capture ? short(params?.messages) : undefined,
          status: "error",
          errorMessage: err instanceof Error ? err.message.slice(0, 4_000) : String(err).slice(0, 4_000),
          latencyMs: Date.now() - t0,
          startedAt,
          endedAt: new Date().toISOString(),
        });
      } catch {
        /* fail-open */
      }
    }
  }) as T["messages"]["create"];
  return client;
}

/* ── OpenAI ────────────────────────────────────────────────────── */

interface OpenAILike {
  chat: { completions: { create: (params: any, ...rest: any[]) => Promise<any> } };
}

/**
 * Wrap an OpenAI SDK client: `instrumentOpenAI(openai, revalex)`.
 * Every `chat.completions.create` call is traced.
 */
export function instrumentOpenAI<T extends OpenAILike>(
  client: T,
  revalex: RevalexClient,
  opts: InstrumentOptions = {},
): T {
  const original = client.chat.completions.create.bind(client.chat.completions);
  const capture = opts.captureContent ?? true;

  client.chat.completions.create = ((params: any, ...rest: any[]) => {
    const startedAt = new Date().toISOString();
    const t0 = Date.now();
    let call: Promise<any>;
    try {
      call = original(params, ...rest);
    } catch (err) {
      recordFailure(err);
      throw err;
    }
    // See the Anthropic wrapper: observe, don't replace — the APIPromise
    // the caller gets back keeps .withResponse()/.asResponse().
    void Promise.resolve(call).then(recordSuccess, recordFailure);
    return call;

    function recordSuccess(result: any): void {
      try {
        const streaming = params?.stream === true;
        const textOut = streaming ? undefined : result?.choices?.[0]?.message?.content ?? undefined;
        revalex.trace({
          type: "llm_call",
          name: opts.name ?? "openai.chat.completions.create",
          model: result?.model ?? params?.model,
          input: capture ? short(params?.messages) : undefined,
          output: capture && textOut ? short(textOut) : undefined,
          tokensIn: result?.usage?.prompt_tokens,
          tokensOut: result?.usage?.completion_tokens,
          latencyMs: Date.now() - t0,
          status: "ok",
          metadata: streaming ? { streaming: true } : undefined,
          startedAt,
          endedAt: new Date().toISOString(),
        });
      } catch {
        /* fail-open */
      }
    }

    function recordFailure(err: unknown): void {
      try {
        revalex.trace({
          type: "llm_call",
          name: opts.name ?? "openai.chat.completions.create",
          model: params?.model,
          input: capture ? short(params?.messages) : undefined,
          status: "error",
          errorMessage: err instanceof Error ? err.message.slice(0, 4_000) : String(err).slice(0, 4_000),
          latencyMs: Date.now() - t0,
          startedAt,
          endedAt: new Date().toISOString(),
        });
      } catch {
        /* fail-open */
      }
    }
  }) as T["chat"]["completions"]["create"];
  return client;
}

File: python/revalex/instrument.py
"""Auto-instrumentation for Anthropic and OpenAI clients.

Duck-typed (zero dependencies): pass the provider client in, its
create() method gets wrapped so every call is traced. Behavior is
never changed — results and errors pass through untouched; capture
failures are swallowed (fail-open).

    from revalex import RevalexClient, instrument_anthropic
    import anthropic

    revalex = RevalexClient(api_key="rvx_live_...")
    client = instrument_anthropic(anthropic.Anthropic(), revalex)
    # every client.messages.create(...) is now traced
"""

from __future__ import annotations

import json
import time
from typing import Any, Optional

from .client import RevalexClient

_MAX_CAPTURE = 20_000


def _short(v: Any) -> Optional[str]:
    try:
        s = v if isinstance(v, str) else json.dumps(v, default=str)
        return s[:_MAX_CAPTURE] + "…" if len(s) > _MAX_CAPTURE else s
    except Exception:
        return None


def instrument_anthropic(client: Any, revalex: RevalexClient, name: Optional[str] = None, capture_content: bool = True) -> Any:
    """Wrap an Anthropic SDK client's messages.create with tracing."""
    original = client.messages.create
    trace_name = name or "anthropic.messages.create"

    def wrapped(*args: Any, **kwargs: Any) -> Any:
        started = time.time()
        try:
            result = original(*args, **kwargs)
        except Exception as e:
            _safe_trace(
                revalex,
                name=trace_name,
                model=kwargs.get("model"),
                input=_short(kwargs.get("messages")) if capture_content else None,
                status="error",
                error_message=str(e)[:4000],
                latency_ms=(time.time() - started) * 1000,
            )
            raise
        try:
            streaming = kwargs.get("stream") is True
            text_out = None
            if not streaming:
                blocks = getattr(result, "content", None) or []
                parts = [getattr(b, "text", None) for b in blocks if getattr(b, "type", None) == "text"]
                text_out = "\n".join(p for p in parts if p) or None
            usage = getattr(result, "usage", None)
            _safe_trace(
                revalex,
                name=trace_name,
                model=getattr(result, "model", None) or kwargs.get("model"),
                input=_short(kwargs.get("messages")) if capture_content else None,
                output=_short(text_out) if (capture_content and text_out) else None,
                tokens_in=getattr(usage, "input_tokens", None),
                tokens_out=getattr(usage, "output_tokens", None),
                status="ok",
                metadata={"streaming": True} if streaming else None,
                latency_ms=(time.time() - started) * 1000,
            )
        except Exception:
            pass  # capture must never break the app
        return result

    client.messages.create = wrapped
    return client


def instrument_openai(client: Any, revalex: RevalexClient, name: Optional[str] = None, capture_content: bool = True) -> Any:
    """Wrap an OpenAI SDK client's chat.completions.create with tracing."""
    original = client.chat.completions.create
    trace_name = name or "openai.chat.completions.create"

    def wrapped(*args: Any, **kwargs: Any) -> Any:
        started = time.time()
        try:
            result = original(*args, **kwargs)
        except Exception as e:
            _safe_trace(
                revalex,
                name=trace_name,
                model=kwargs.get("model"),
                input=_short(kwargs.get("messages")) if capture_content else None,
                status="error",
                error_message=str(e)[:4000],
                latency_ms=(time.time() - started) * 1000,
            )
            raise
        try:
            streaming = kwargs.get("stream") is True
            text_out = None
            if not streaming:
                choices = getattr(result, "choices", None) or []
                if choices:
                    message = getattr(choices[0], "message", None)
                    text_out = getattr(message, "content", None)
            usage = getattr(result, "usage", None)
            _safe_trace(
                revalex,
                name=trace_name,
                model=getattr(result, "model", None) or kwargs.get("model"),
                input=_short(kwargs.get("messages")) if capture_content else None,
                output=_short(text_out) if (capture_content and text_out) else None,
                tokens_in=getattr(usage, "prompt_tokens", None),
                tokens_out=getattr(usage, "completion_tokens", None),
                status="ok",
                metadata={"streaming": True} if streaming else None,
                latency_ms=(time.time() - started) * 1000,
            )
        except Exception:
            pass
        return result

    client.chat.completions.create = wrapped
    return client


def _safe_trace(revalex: RevalexClient, **kwargs: Any) -> None:
    try:
        revalex.trace(type="llm_call", **{k: v for k, v in kwargs.items() if v is not None})
    except Exception:
        pass
